### DISCUSSION TITLE
[3.0.7] Restricting Tailwind variants selectors to Tailwind classes makes it difficult to work with external libraries

### DISCUSSION THREAD

**Discussion by SadiinsoSnowfall** [tailwindlabs/tailwindcss#8156] [Help] [answered]
Hi,
In one of my project using Angular Material 13 and Tailwindcss 3 I have the following scss rule:
```scss
.test { @apply hover:mat-elevation-z4; }
```
`mat-elevation-z4` being a class imported from Angular/material. This does work when using tailwindcss<=3.0.6 but stopped working on newer version, instead I get this error:
```
Error: Module build failed (from ./node_modules/postcss-loader/dist/cjs.js):
SyntaxError

(317:1) styles.scss The `hover:mat-elevation-z4` class does not exist. If `hover:mat-elevation-z4` is a custom class, make sure it is defined within a `@layer` directive.

  315 | .mat-form-field-appearance-legacy .mat-form-field-subscript-wrapper {
  316 |   margin-top: 0.5416666667em;
> 317 |   top: calc(100% - 1.6666666667em);
      | ^
  318 | }
  319 | 

    at Object.<anonymous> (node_modules\css-loader\dist\cjs.js??ruleSet[1].rules[4].rules[0].oneOf[1].use[1]!node_modules\postcss-loader\dist\cjs.js??ruleSet[1].rules[4].rules[0].oneOf[1].use[2]!node_modules\resolve-url-loader\index.js??ruleSet[1].rules[4].rules[1].use[0]!node_modules\sass-loader\dist\cjs.js??ruleSet[1].rules[4].rules[1].use[1]!front\src\styles.scss:1:7)
    at node_modules\webpack\lib\javascript\JavascriptModulesPlugin.js:432:11
    at Hook.eval [as call] (eval at create (node_modules\tapable\lib\HookCodeFactory.js:19:10), <anonymous>:7:1)
    at Hook.CALL_DELEGATE [as _call] (node_modules\tapable\lib\Hook.js:14:14)
    at node_modules\webpack\lib\Compilation.js:4981:39
    at tryRunOrWebpackError (node_modules\webpack\lib\HookWebpackError.js:83:7)
    at __webpack_require_module__ (node_modules\webpack\lib\Compilation.js:4979:12)
    at __webpack_require__ (node_modules\webpack\lib\Compilation.js:4936:18)
```
(the code displayed in the error is from angular material, not my scss file)
I tried this same rule on a fresh angular project to see if it was because I missconfigured something but I still get the same result.

I also observed the following behaviour:
```scss
.test { @apply hover:mat-elevation-z4; } // doesn't compile when using tailwindcss>=3.0.7, subject of this discussion
.test { @apply mat-elevation-z4; } // always compile & works
.test { @apply hover:bg-blue-500; } // always compile & works
```
```html
<p class="mat-elevation-z4">test</p> <!-- always works  -->
<p class="hover:mat-elevation-z4">test</p> <!-- is ignored when using tailwindcss>=3.0.7  -->
```
Is there a reason as to why this doesn't work with newer tailwind versions ?  I couldn't find anything online or on this github repo about this issue.

**Comment by SadiinsoSnowfall** [comment]
The dependencies of my project are the following :
```
  "dependencies": {
    "@angular/animations": "~13.0.0",
    "@angular/cdk": "^13.3.3",
    "@angular/common": "~13.0.0",
    "@angular/compiler": "~13.0.0",
    "@angular/core": "~13.0.0",
    "@angular/forms": "~13.0.0",
    "@angular/material": "^13.3.3",
    "@angular/platform-browser": "~13.0.0",
    "@angular/platform-browser-dynamic": "~13.0.0",
    "@angular/router": "~13.0.0",
    "@fortawesome/angular-fontawesome": "^0.10.2",
    "@fortawesome/fontawesome-svg-core": "^6.1.0",
    "@fortawesome/free-brands-svg-icons": "^6.1.0",
    "@fortawesome/free-regular-svg-icons": "^6.1.0",
    "@fortawesome/free-solid-svg-icons": "^6.1.0",
    "rxjs": "~7.4.0",
    "tslib": "^2.3.0",
    "zone.js": "~0.11.4"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~13.0.2",
    "@angular/cli": "~13.0.2",
    "@angular/compiler-cli": "~13.0.0",
    "@types/jasmine": "~3.10.0",
    "autoprefixer": "^10.4.4",
    "jasmine-core": "~3.10.0",
    "karma": "~6.3.0",
    "karma-chrome-launcher": "~3.1.0",
    "karma-coverage": "~2.0.3",
    "karma-jasmine": "~4.0.0",
    "karma-jasmine-html-reporter": "~1.7.0",
    "postcss": "^8.4.12",
    "tailwindcss": "^3.0.24",
    "typescript": "~4.4.3"
  }
  ```
  I am not using any tailwindcss plugin.

**Comment by SadiinsoSnowfall** [comment]
Ok I found #6589 which disallowed exactly what I describe in this discussion.

This mean that since version 3.0.7 the following doesn't work :
```scss
.a-class { color: red; }
.test { @apply hover:a-class; }
```
Whereas the following does work:
```scss
@layer utilities {
    .a-class { color: red; }
}
.test { @apply hover:a-class; }
```
The problem is that I can't add a `@layer` directive to the `mat-elevation-z4` class, as it is defined in the angular material sources.

I can do something like this to work around the problem :
```scss
@layer utilities {
    .proxy {
        @apply mat-elevation-z4;
    }
}

.test {
    @apply hover:proxy;
}
// also works for <div class="hover:proxy"></div>
```
But this seems a bit too verbose, and would require me to create a proxy class for every custom class from external libraries I whish to use.

**Comment by Serenit8** [comment]
These classes follow the pattern mat-elevation-z# , where # is the elevation number you want, from 0 to 24. These predefined classes use the CSS box-shadow.
[LTD Commodities](https://www.ltdcommodities.me/)

**Comment by SadiinsoSnowfall** [comment]
Yes indeed, but I was not looking to understand how mat-elevations helpers work in Angular, but rather to point out that restricting the classes accepted by Tailwind variants to Tailwind classes also exclude the classes from other libraries (Angular Material for example) and makes it more difficult to work with such libs.

**Comment by SadiinsoSnowfall** [comment]
I have updated the discussion title to reflect this

**Comment by adamwathan** [comment, marked as answer]
Hey! I would recommend just creating a custom class that does what you need but lives in a layer so it can be used with apply. No real plans to support this I’m afraid or double down on apply anymore than we already have, I hate the feature and regret adding it to the framework because of how many support requests and bug reports it’s caused 😫

If you’re using apply more than 5 or 6 times in your project you aren’t really using Tailwind the way it was intended to be used, and we’re only interested in developing the framework for the recommended workflow.
